examples/with-apollo-auth/lib/apollo.js

```diff
@@ -0,0 +1,195 @@
+import React, { useMemo } from 'react'
+import PropTypes from 'prop-types'
+import cookie from 'cookie'
+import { ApolloProvider } from '@apollo/react-hooks'
+import Head from 'next/head'
+import { ApolloClient, InMemoryCache, HttpLink } from 'apollo-boost'
+import { setContext } from 'apollo-link-context'
+import fetch from 'isomorphic-unfetch'
+
+/**
+ * Creates and provides the apolloContext
+ * to a next.js PageTree. Use it by wrapping
+ * your PageComponent via HOC pattern.
+ * @param {Function|Class} PageComponent
+ * @param {Object} [config]
+ * @param {Boolean} [config.ssr=true]
+ */
+export function withApollo (PageComponent) {
+  const WithApollo = ({ apolloClient, apolloState, ...pageProps }) => {
+    const client = useMemo(() => {
+      // We pass in the apolloClient directly when using getDataFromTree
+      if (apolloClient) {
+        return apolloClient
+      }
+
+      // Otherwise initClient using apolloState
+      return initApolloClient(apolloState, {
+        getToken: () => {
+          return parseCookies().token
+        }
+      })
+    }, [])
+    return (
+      <ApolloProvider client={client}>
+        <PageComponent {...pageProps} />
+      </ApolloProvider>
+    )
+  }
+
+  if (process.env.NODE_ENV !== 'production') {
+    // Find correct display name
+    const displayName =
+      PageComponent.displayName || PageComponent.name || 'Component'
+
+    // Warn if old way of installing apollo is used
+    if (displayName === 'App') {
+      console.warn('This withApollo HOC only works with PageComponents.')
```

Comment:
Thank you @HaNdTriX for all your work.
 
What will happen if I still keep my `_app.js` and use
```
export default withApollo(MyApp);
```

Reply:
This won't work since the types of the params of `getInitialProps` and `appProps` differs in both use cases.

`App.getInitialProps<appContext > !== Page.getInitialProps<pageContext>`
`App<appProps> !== Page<pageProps>`

Please note that it would be possible to change the HOC to work with the App.
Nevertheless, we don't recommend this pattern since you will disable automatic static optimization for every route in your Next.js App.

Reply:
Thank you it's a lot clearer now.
It's sad because all of my pages use Apollo. So it's cumbersome to add that everywhere now

Reply:
You can still use the old implementation.

Reply:
@HaNdTriX maybe we should add another example using `_app.js`? Most "apps" that don't have a lot of static pages will most likely find using `_app.js` easier with less boilerplate.